Feature/full outer join

kaelte opened the pull request:
temporalFullJoin, temporalLeftJoin and temporalRightJoin use
```
private def temporalKeyOuterJoinImpl( df1:DataFrame, df2:DataFrame, keys:Seq[String], rnkExpressions:Seq[Column], additionalJoinFilterCondition:Column, joinType:String = "full" )
                                     (implicit ss:SparkSession, hc:TemporalQueryConfig): DataFrame
```

kaelte committed:
temporalFullJoin and temporalRightJoin implemented

kaelte committed:
type annotation added, English improved

zzeekk commented on src/main/scala/ch/zzeekk/spark/temporalquery/TemporalQueryUtil.scala:
```diff
@@ -159,16 +186,27 @@ object TemporalQueryUtil extends LazyLogging {
    */
   private def temporalJoinImpl( df1:DataFrame, df2:DataFrame, keyCondition:Column, joinType:String = "inner" )(implicit ss:SparkSession, hc:TemporalQueryConfig) : DataFrame = {
     // history join
+    df1.orderBy("id",hc.fromColName).show(false)
```
Ist das noch debugging?

kaelte replied:
Yes :blush: Removed: https://github.com/zzeekk/spark-temporalquery/pull/16/commits/66dcef077bca093ab14dd082ea60fbb169398d3d

zzeekk commented on src/main/scala/ch/zzeekk/spark/temporalquery/TemporalQueryUtil.scala:
```diff
@@ -159,16 +186,27 @@ object TemporalQueryUtil extends LazyLogging {
+    df1.orderBy("id",hc.fromColName).show(false)
+
     val df2ren = df2.withColumnRenamed(hc.fromColName,hc.fromColName2).withColumnRenamed(hc.toColName,hc.toColName2)
+    df2ren.orderBy("id",hc.fromColName2).show(false)
```
Ist das noch debugging?

kaelte replied:
Yes :blush: Removed: https://github.com/zzeekk/spark-temporalquery/pull/16/commits/66dcef077bca093ab14dd082ea60fbb169398d3d

zzeekk commented on src/main/scala/ch/zzeekk/spark/temporalquery/TemporalQueryUtil.scala:
```diff
@@ -159,16 +186,27 @@ object TemporalQueryUtil extends LazyLogging {
     val df2ren = df2.withColumnRenamed(hc.fromColName,hc.fromColName2).withColumnRenamed(hc.toColName,hc.toColName2)
+    df2ren.orderBy("id",hc.fromColName2).show(false)
     val df_join = df1.join( df2ren, keyCondition and col(hc.fromColName)<=col(hc.toColName2) and col(hc.toColName)>=col(hc.fromColName2), joinType)
+    df_join.show(false)
```
Ist das noch debugging?

kaelte replied:
Yes :blush: Removed: https://github.com/zzeekk/spark-temporalquery/pull/16/commits/66dcef077bca093ab14dd082ea60fbb169398d3d

zzeekk commented on src/main/scala/ch/zzeekk/spark/temporalquery/TemporalQueryUtil.scala:
```diff
@@ -159,16 +186,27 @@ object TemporalQueryUtil extends LazyLogging {
-    val df2Cols = df2.columns.diff(df1Cols ++ hc.technicalColNames)
-    val selCols: Array[Column] = df1Cols.map(df1(_)) ++ df2Cols.map(df2(_)) :+ greatest(col(hc.fromColName), col(hc.fromColName2)).as(hc.fromColName) :+ least(col(hc.toColName), col(hc.toColName2)).as(hc.toColName)
-    df_join.select(selCols:_*)
+    val dfCommonColNames = df1.columns.intersect(df2.columns).diff(hc.technicalColNames)
```
What is the improvement of this refactoring?

kaelte replied:
With the old implementation, 2 test cases fail: temporalFullJoin_dfRight, temporalRightJoin_dfRight.

The reason is that the old implementation gives priority to the columns of df₁: common columns of df₁ and df₂ are selected from df₁ only as they are removed from df2.columns:

```
val df1Cols = df1.columns.diff(hc.technicalColNames)
val df2Cols = df2.columns.diff(df1Cols ++ hc.technicalColNames)

val selCols: Array[Column] = df1Cols.map(df1(_)) ++ df2Cols.map(df2(_)) :+ ...
```

Note that the common columns occur in df1Cols only, but not in df2Cols.

In case you want to use `temporalJoinImpl` for a right join you would need to do the opposite:
```
val df2Cols = df2.columns.diff(hc.technicalColNames)
val df1Cols = df1.columns.diff(df2Cols ++ hc.technicalColNames)

val selCols: Array[Column] = df1Cols.map(df1(_)) ++ df2Cols.map(df2(_))) :+ ...
```

And in case of a full join ?

Well, best is an almost symmetric implementation where we select the common columns using `coalesce` as done now.

Why do only these 2 test cases fail?

Because these are the only tests where an entity occurs in df₂ only: There is not record with `id=1` in df₁.

zzeekk replied:
Sounds good, thanks for the rich explanation.

zzeekk commented on src/main/scala/ch/zzeekk/spark/temporalquery/TemporalQueryUtil.scala:
```diff
@@ -225,20 +263,21 @@ object TemporalQueryUtil extends LazyLogging {
    */
-  private def temporalKeyLeftJoinImpl( df1:DataFrame, df2:DataFrame, keys:Seq[String], rnkExpressions:Seq[Column], additionalJoinFilterCondition:Column )
-                                     (implicit ss:SparkSession, hc:TemporalQueryConfig) = {
+  private def temporalKeyOuterJoinImpl( df1:DataFrame, df2:DataFrame, keys:Seq[String], rnkExpressions:Seq[Column], additionalJoinFilterCondition:Column, joinType:String = "full" )
```
```suggestion
  private def temporalKeyOuterJoinImpl( df1:DataFrame, df2:DataFrame, keys:Seq[String], rnkExpressions:Seq[Column], additionalJoinFilterCondition:Column, joinType:String)
```
I suggest to not set a default value forjoinType. When using this function one should be forced to choose the join type.

kaelte replied:
default value for joinType removed: https://github.com/zzeekk/spark-temporalquery/pull/16/commits/f7f3f48cc316eb731c30380803c2dd5979161fb1

zzeekk commented on src/main/scala/ch/zzeekk/spark/temporalquery/TemporalQueryUtil.scala:
```diff
@@ -225,20 +263,21 @@ object TemporalQueryUtil extends LazyLogging {
   /**
    * left outer join
```
```suggestion
   * outer join
```

kaelte replied:
Word removed: https://github.com/zzeekk/spark-temporalquery/pull/16/commits/f7f3f48cc316eb731c30380803c2dd5979161fb1

zzeekk reviewed (commented):
Hi @kaelte , looks good in general, but i have some questions to be clarified.

kaelte committed:
forgotten debugging shows removed :)

kaelte committed:
temporalKeyOuterJoinImpl: default value for joinType removed

When using temporalKeyOuterJoinImpl one should be forced to choose the join type.

kaelte committed:
commented code removed
